File: src/model-stats-client.js
'use strict';

// Thin wrapper around model-stats-db.js's CLI (record-call / record-outcome), letting
// local-draft.js/review-task.js/reject-retry-check.js call it as a normal async function
// instead of each hand-rolling its own temp-file-plus-execFileSync boilerplate.
// model-stats-db.js itself can't be require()'d directly -- it runs its logic
// unconditionally at module-load time (no `require.main === module` guard) and calls
// process.exit() on error, so it has to be invoked as a real subprocess, matching how the
// reference (Invoke-ModelStatsDb in ornith-worker.ps1/review-runner.ps1/queue-watchdog.ps1)
// always shelled out to it too.
//
// Best-effort by design: stats tracking must never break real pipeline work over a
// recording failure (a locked db file, a bad payload) -- every function here swallows its
// own errors, same "log-and-continue" treatment the reference gives this exact call.

const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');
const { estimateApiCostUsd } = require('./anthropic-pricing.js');

const SCRIPT_PATH = path.join(__dirname, 'model-stats-db.js');

function runEvent(event, payload) {
  const tmpPath = path.join(os.tmpdir(), `model-stats-${event}-${Date.now()}-${Math.random().toString(36).slice(2)}.json`);
  try {
    fs.writeFileSync(tmpPath, JSON.stringify(payload));
    // --no-warnings: node:sqlite (model-stats-db.js's own DB layer, migrated off
    // better-sqlite3 2026-08-19) emits an ExperimentalWarning on every single invocation
    // otherwise -- harmless, but this call already runs with stdio:'pipe' specifically to
    // stay quiet, and the warning would otherwise land in whatever captures this process's
    // stderr (e.g. local-worker.sh's own log redirect) on every implement/review pass.
    execFileSync('node', ['--no-warnings', SCRIPT_PATH, event, tmpPath], { stdio: 'pipe' });
  } catch (e) {
    // Non-fatal -- see header.
  } finally {
    try { fs.unlinkSync(tmpPath); } catch (e) { /* best-effort cleanup */ }
  }
}

// Call this once per implement-pass model call, right after it completes -- returns a
// fresh callId to store on the task (as task.abCallId) so a later outcome (review verdict,
// watchdog requeue) can be joined back to this same row.
//
// candidates (2026-08-19): the raw LOCAL_AB_MODELS list this call was chosen from, or
// null when no A/B mechanism is active for this call -- model-stats-db.js's schema already
// had this column (the PowerShell reference always populated it), it just had no way to
// reach it through this JS wrapper until ab-model-select.js gave local-draft.js a real
// selection to record.
function recordCall({ taskId, stage = 'implement', model, candidates = null, startedAt, latencyMs, result }) {
  const callId = require('crypto').randomUUID();
  runEvent('record-call', {
    callId,
    taskId,
    stage,
    model,
    candidates,
    startedAt,
    latencyMs,
    evalDurationNs: result && result.eval_duration != null ? result.eval_duration : null,
    promptEvalCount: result && result.prompt_eval_count != null ? result.prompt_eval_count : null,
    evalCount: result && result.eval_count != null ? result.eval_count : null,
    // instanceId (2026-08-23, "Where else would it make sense to track it?" -> Workers
    // tab, per-instance cumulative cost): AGENT_MANAGER_INSTANCE_ID is already a real env
    // var in every worker's process tree (local-worker.sh exports it -- see
    // model-inflight-lock.js's own use of it for the exact same kind of attribution), read
    // directly here rather than threading a new param through all 3 call sites.
    instanceId: process.env.AGENT_MANAGER_INSTANCE_ID || null,
    // costUsd (2026-08-23, Grimmethy: "Do we have any way of knowing how much these
    // tasks would cost using anthropic API?"): claude-client.js's call() has always
    // computed this (Claude Code CLI's own total_cost_usd, a client-side estimate
    // against real Anthropic API pricing, independent of subscription billing) -- it
    // just never reached this far. Undefined/null for every Ornith result (that
    // module's result shape has no costUsd field at all), which is exactly what should
    // happen: a null cost_usd row means "free local call," not "unknown cost."
    costUsd: result && result.costUsd != null ? result.costUsd : null,
    // hypotheticalCostUsd (2026-08-23, Grimmethy: "Clarification on the anthropic costs.
    // I'd like estimates for if we had used the API. Even if we used the local models.")
    // -- costUsd above is real spend for an ACTUAL Claude call and null for a local one;
    // this is "what would THIS call have cost on the API regardless of what actually ran
    // it" -- the real costUsd when this genuinely was a Claude call (no token-count
    // fields exist on that result shape to estimate from, and the real number is already
    // exact), or a token-based estimate (anthropic-pricing.js, against real
    // prompt_eval_count/eval_count Ollama already reports) for a local call. Always
    // populated (never null) so SUM(hypothetical_cost_usd) alone answers "what if
    // everything this period had gone through the API," while SUM(cost_usd) still
    // answers "what did we actually spend" -- two different, both real, questions.
    hypotheticalCostUsd: result && result.costUsd != null
      ? result.costUsd
      : estimateApiCostUsd({
        promptTokens: result && result.prompt_eval_count != null ? result.prompt_eval_count : 0,
        completionTokens: result && result.eval_count != null ? result.eval_count : 0,
      }),
    attempts: result && result.attempts != null ? result.attempts : null,
    degenerate: result && result.degenerate != null ? result.degenerate : null,
  });
  return callId;
}

function recordOutcome({ callId, outcome, outcomeStage, outcomeReason }) {
  if (!callId) return; // matches the reference: no callId (e.g. a pre-drafted task never had an implement-pass call) means nothing to update.
  runEvent('record-outcome', { callId, outcome, outcomeStage, outcomeReason });
}

// Reads back the running Anthropic-API-equivalent cost total (2026-08-23, see
// recordCall's own costUsd comment). Unlike runEvent() above (fire-and-forget,
// stdio:'pipe', built for record-call/record-outcome's write-only shape), this
// captures and parses real stdout -- model-stats-db.js's own `cost-summary` event
// prints JSON: { totalCostUsd, callsWithCost, freeCalls, byModel: [...], byDay: [...] }.
// Same best-effort philosophy as the rest of this file: returns null on any failure
// (missing db, no calls recorded yet) rather than throwing -- a cost dashboard widget
// failing to render is not worth breaking whatever page/report is asking for it.
function getCostSummary() {
  try {
    const stdout = execFileSync('node', ['--no-warnings', SCRIPT_PATH, 'cost-summary'], { stdio: ['ignore', 'pipe', 'pipe'], encoding: 'utf8' });
    return JSON.parse(stdout);
  } catch (e) {
    return null;
  }
}

module.exports = { recordCall, recordOutcome, getCostSummary };

File: python/dashboard/model_stats_client.py
"""Python-side wrapper for src/model-stats-db.js's record-call/record-outcome CLI --
the Python equivalent of src/model-stats-client.js, so the dashboard's interactive
sessions (Discuss, Grill Me/Grill With Docs) show up in the same model-stats.db the
Models tab already reads, on both providers. Before this, only the Node pipeline's
implement-pass calls were ever recorded -- every interactive dashboard session (Ollama
or, now, Claude) was invisible to the Models tab regardless of provider.

Same "never let stats recording break the real feature" contract as the JS version:
every function here swallows its own errors.
"""
import json
import logging
import subprocess
import tempfile
import uuid
from datetime import datetime, timezone
from pathlib import Path

SRC_DIR = Path(__file__).resolve().parent.parent.parent / "src"
MODEL_STATS_DB_JS = SRC_DIR / "model-stats-db.js"

logger = logging.getLogger(__name__)


def _run_event(event: str, payload: dict):
    tmp_path = Path(tempfile.gettempdir()) / f"model-stats-{event}-{uuid.uuid4().hex}.json"
    try:
        tmp_path.write_text(json.dumps(payload), encoding="utf-8")
        subprocess.run(["node", str(MODEL_STATS_DB_JS), event, str(tmp_path)],
                        capture_output=True, timeout=15)
    except (OSError, subprocess.SubprocessError):
        pass
    finally:
        try:
            tmp_path.unlink()
        except OSError:
            pass


def record_call(task_id: str, model: str, latency_ms: int, stage: str = "discuss",
                 result: dict = None, started_at: str = None, source: str = None) -> str:
    """Call once per interactive-session model call. Returns a callId for a later
    record_outcome() (there usually isn't one for free-form chat -- discuss/grill
    sessions have no approve/reject verdict -- so most callers just keep the callId
    unused; it's still useful as a stable row identifier if a caller wants it).

    turnsUsed/source (2026-08-26, Grimmethy: "add turnsUsed recording... a data point we
    track for each job type in the Job List itself (min/max/average)"): chat_sessions.py's
    own local_tool_client.stream_plan_with_tools() result always carries a real turnsUsed
    count -- it just never reached this far before. result.get("turnsUsed") is None for
    every shape that isn't a runPlanWithTools()-backed result (e.g. a bare Claude Code
    reply), matching this file's own existing degenerate-field convention."""
    call_id = str(uuid.uuid4())
    result = result or {}
    _run_event("record-call", {
        "callId": call_id,
        "taskId": task_id,
        "stage": stage,
        "model": model,
        "startedAt": started_at or datetime.now(timezone.utc).isoformat(),
        "latencyMs": latency_ms,
        "evalDurationNs": None,
        "promptEvalCount": None,
        "evalCount": None,
        "attempts": None,
        "degenerate": result.get("degenerate"),
        "turnsUsed": result.get("turnsUsed"),
        "source": source,
    })
    return call_id


def get_turns_summary() -> dict | None:
    """Per-source MIN/MAX/AVG(turns_used) (2026-08-26, see record_call's own turnsUsed
    comment) -- for the Job List tab. Same best-effort-returns-None shape as the JS
    wrapper's getTurnsSummary(), captures real stdout instead of the fire-and-forget
    _run_event() pattern above since this actually needs the result."""
    try:
        proc = subprocess.run(
            ["node", "--no-warnings", str(MODEL_STATS_DB_JS), "turns-summary"],
            capture_output=True, text=True, timeout=15,
        )
        if proc.returncode != 0:
            return None
        return json.loads(proc.stdout)
    except (OSError, subprocess.SubprocessError, json.JSONDecodeError) as exc:
        logger.warning("model-stats client failed: %s", exc, exc_info=True)
        return None


def record_outcome(call_id: str, outcome: str, outcome_stage: str = "discuss", outcome_reason: str = None):
    if not call_id:
        return
    _run_event("record-outcome", {
        "callId": call_id,
        "outcome": outcome,
        "outcomeStage": outcome_stage,
        "outcomeReason": outcome_reason,
    })
